clang/tools/clang-offload-deps/ClangOffloadDeps.cpp

```diff
@@ -0,0 +1,254 @@
+      "A tool for creating dependence bitcode files for offload targets. "
+      "Takes\nhost image as input and produces bitcode files, one per offload "
+      "target, with\nreferences to symbols that must be defined in target "
+      "images.\n");
+
+  if (Help) {
+    cl::PrintHelpMessage();
+    return 0;
+  }
+
+  // The number of output files and targets should match.
+  if (Targets.size() != Outputs.size()) {
+    reportError(
+        createStringError(errc::invalid_argument,
+                          "number of output files and targets should match"));
+    return 1;
+  }
+
+  // Verify that given targets are valid. Each target string is expected to have
+  // the following format
+  //     <kind>-<triple>
+  // where <kind> is host, openmp, hip, sycl or fpga,
+  // and <triple> is an offload target triple.
+  SmallVector<StringRef, 8u> Triples(Targets.size());
+  for (unsigned I = 0; I < Targets.size(); ++I) {
+    StringRef Kind;
+    std::tie(Kind, Triples[I]) = StringRef(Targets[I]).split('-');
+
+    bool KindIsValid = !Kind.empty() && StringSwitch<bool>(Kind)
+                                            .Case("host", true)
+                                            .Case("openmp", true)
+                                            .Case("hip", true)
+                                            .Case("sycl", true)
+                                            .Case("fpga", true)
+                                            .Default(false);
+
+    bool TripleIsValid = !Triples[I].empty() &&
+                         Triple(Triples[I]).getArch() != Triple::UnknownArch;
+
+    if (!KindIsValid || !TripleIsValid) {
+      SmallVector<char, 128u> Buf;
+      raw_svector_ostream Msg(Buf);
+      Msg << "invalid target '" << Targets[I] << "'";
+      if (!KindIsValid)
+        Msg << ", unknown offloading kind '" << Kind << "'";
+      if (!TripleIsValid)
+        Msg << ", unknown target triple '" << Triples[I] << "'";
+      reportError(createStringError(errc::invalid_argument, Msg.str()));
+      return 1;
+    }
+  }
+
+  // Read input file. It should have one of the supported object file formats.
+  Expected<OwningBinary<ObjectFile>> ObjectOrErr =
+      ObjectFile::createObjectFile(Input);
+  if (!ObjectOrErr) {
+    reportError(ObjectOrErr.takeError());
+    return 1;
+  }
+
+  // Then try to find a section in the input binary which contains offload
+  // symbol names and parse section contents.
+  DenseMap<StringRef, SmallDenseSet<StringRef>> Target2Symbols;
+  for (SectionRef Section : ObjectOrErr->getBinary()->sections()) {
+    // Look for the .tgtsym section in the binary.
+    Expected<StringRef> NameOrErr = Section.getName();
+    if (!NameOrErr) {
+      reportError(NameOrErr.takeError());
+      return 1;
+    }
+    if (*NameOrErr != SYMBOLS_SECTION_NAME)
+      continue;
+
+    // This is the section we are looking for, read symbol names from it.
+    Expected<StringRef> DataOrErr = Section.getContents();
```

Comment:
Is it expected that we consider files w/o `.tgtsym` as bad input? Will it be a problem to produce an empty list of dependencies for such files?

Reply:
Such inputs are not expected to cause any errors. I will update LIT test to check that we produce an empty list for such inputs.

Reply:
Right... misread the code. I see that error is reported for empty `.tgtsym` sections.